### ISSUE TITLE
Vertaisarviointi

### ISSUE THREAD

**Issue by JuusoSaavalainen** [luukasmakila/neural-network-to-recognize-digits#1]
## Projekti ladattu ja testattu klo 13.00 19.2.2023

Koodi pääosin selkeää ja suoraviivaista. Monet asiat ovat totetuttu todella elegantisti eikä projektista ollut helppo löytää parannus ehdotuksia. Asentaminen ja testaus onnistui helposti.  Ekstra pointit koodin luettavuudesta sekä yksinkertaisesta rakentaasta läpi projektin.

Tässä kuitenkin joitain asioita joita jäin miettimään projektia tutkiessani ja joissa mielestäni on vielä parantamisen varaa.

### 1. Vanishing gradient problem
 
Malli tuottaa hienoja tuloksia annetuilla parametreillä. Koska malli on rakennettu modulaariseksi 

```
# Activation functions 
def sigmoid(z_vector):
    """
    Sigmoid activation function
    """
    return (1.0/(1.0 + np.exp(-z_vector)))

def sigmoid_prime(z_vector):
    """
    Sigmoid functions derivative
    """
    return sigmoid(z_vector)*(1-sigmoid(z_vector)) 
```

Pelkästään näitä aktivaatio functioita käyttämällä kun layereitä lisätään saatetaan törmätä katoavan gradientin ongelmaan.
Ehkä voisit käyttää myös muita aktivaatio fuktioita? Ainakin jos on tarkoituksena että mallia voi muokata olisi hyvä sisällyttää työkaluja jotka varmistavat toimivuuden myös tilanteissa joissa niin tehdään.

### 2.  Algoritmien riippuvuudet

Backpropagation rakennettu feedforwardista riippuvaiseksi. En keksinyt tälle mitään järkevää syytä. 
```
activations, z_vectors = self.forwardpropagation(x)
```

Datan sisään ottoon on täten myös kaksi functioita;

```
def feedforward(self, x):
```

```
def forwardpropagation(self, x):
```
Luulen että yhdellä voitaisiin pärjätä ja koodin toisteisuutta saataisiin myös samalla vähennettyä . Nykyinen vaikuttaa käyttävän toista trainingiin ja toista test datan testaamiseen [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) .  

### 3.  Visualisointi ja monitorointi

Training loopin aikana ajataan jokaisen täyden training kierroksen jälkeen mallin läpi testaus dataa. Tämä hidastaa trainingia. 
Mallin optimointia ajatellen voisi olla hyvä monitoroida jotain muita piirteitä / tapoja käyttämällä esimerkiksi jotain seuraavista;
*Classification Accuracy, Logarithmic Loss,Confusion Matrix, F1 Score,Mean Absolute Error, Mean Squared Error*
Toki testidatan luokittelu on myös hyvä asia nähdä, se antaa kuitenkin kovin rajallisen määrän tietoa kouluttamisen aikana tapahtuvasta oppimisesta.

### 4. Datan lataaminen ja riippuvuudet

Projektia ladatessa huomasin asentelevani yli puolen gigatavun kirjastoa. Onko jokin hyvä syy sille miksi alkuperäisen datan lataaminen on ulkoistettu tensorflowille ?  [MNIST](http://yann.lecun.com/exdb/mnist/) (92.7 MB) ?  Tensorflow (588.3 MB) ? Tensorflowilla ladattava datasetti normalisoidaan myös tf buildin functiolla;
```
x_train = tf.keras.utils.normalize(x_train, axis=1)
x_test = tf.keras.utils.normalize(x_test, axis=1)
```
Projektin kehys huomioiden, nykyinen ratkaisu voi toki muistin kannalta olla parempi , tradeoffina tensorflowin pakoitettu lataus vain tätä varten.
